sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -106,6 +109,28 @@
  *     .withNumSplits(30))
  *
  * }</pre>
+ *
+ * *
+ *
+ * <p>To configure a MongoDB sink and update, you must specify a connection {@code URI}, a {@code
+ * Database} * name, a {@code Collection} name. It matches the key with _id in target collection.
+ * For instance: * *
+ *
+ * <pre>{@code
+ * * pipeline
+ * *   .apply(...)
+ * *   .apply(MongoDbIO.write()
+ * *     .withUri("mongodb://localhost:27017")
+ * *     .withDatabase("my-database")
+ * *     .withCollection("my-collection")
+ * *     .withIsUpdate(true)
+ * *     .withUpdateKey("key-to-match")
+ * *     .withUpdateField("field-to-update")
+ * *     .withUpdateOperator("$set")
+ * *     .withNumSplits(30))
```

Comment:
In this case, only one field can be updated. Is that correct? would it make sense to support updating more than one field?
I see that mongodb works with pairs of `updateOperator`+`updateField`. I wonder if it makes sense to support something like that instead?

e.g.:

```
  pipeline
    .apply(...)
    .apply(MongoDbIO.write()
      .withUri("mongodb://localhost:27017")
      .withDatabase("my-database")
      .withCollection("my-collection")
      .withIsUpdate(true)
      .withUpdateKey("key-to-match")
      .withUpdateFields(
                 "$set", "field1",
                 "$currentDate", "datefield1",
                 "$min", "special-minimum-field1")
      .withNumSplits(30))
```

or perhaps something like this:

```
      .withUpdateFields(
                 UpdateField.of("$set", "field1"),
                 UpdateField.of("$currentDate", "datefield1"),
                 UpdateField.of("$min", "special-minimum-field1")))
```

I also think it may be helpful to support  _upsert_ operations. WDYT?

I don't want to make this too complex - but it feels like these options could be supported easily, right?

Reply:
Hi Pablo. Thanks for the inputs. I thought the same. But I am confused how the actual values come in?  With the single field update whatever value comes in from the pipeline will be set as the value for the field. Now if we have to update for multiple fields how it can be done ? I am not pretty sure how it works in beam. Hope my question is clear. 

Upserts functionality can definitely be added.

Reply:
my guess is that maybe we have an document like this in the database:

```
{
 "name": "ironman",
 "status": "active",
 "age": 55,
 "location": "malibu",
 "lastUpdated": "(SOME TIMESTAMP)"
}
```

And perhaps the PCollection that comes as input contains the following element:

```
{
 "name": "ironman",
 "status": "inactive",
 "age": 56,
 "location": "malibu"
}
```

In this case, we could do:
```
      .withUpdateKey("name")
      .withUpdateFields(
                 UpdateField.of("$set", "status"),
                 UpdateField.of("$currentDate", "lastUpdated"),
                 UpdateField.of("$set", "age")))
```

Does that make sense? Another thing I wonder about is - what if we just want to update all the fields without having to list them all one by one on the `updateFields` attribute of the transform.

thoughts?

Reply:
Thanks for the clarification! This makes sense. But in real time Sync (CDC) scenario mostly it will be rows coming in and being added to an array or set within mongoDB. For eg say we have a customer collection. New transactions made by the customer is captured and it should be added to the transactions array within customer document. So we definitely need an option to update with full document. So I think we can add both of these options - updating all the fields together and one by one as well. That will be more feature rich as well. What say ?

Reply:
and other challenge - right now in your example we are assuming that the field/column name in the source is same as field name to be updated in the target. It can be different in most of the cases. So it might be like below:
.withUpdateKey("name")
      .withUpdateFields(
                 UpdateField.of("$set", "status", "dest-status-field-in-mdb"),
                 UpdateField.of("$currentDate", "lastUpdated", "dest-lastUpdated-field-in-mdb"),
                 UpdateField.of("$set", "age", "dest-age-field-in-mdb")))

Reply:
This makes sense to me! : )

I suppose when we update the full document, then the field names must be the same?

Reply:
just for my information - is this something that you'll want to take on?

Reply:
Yeah. Have implemented most of it. Testing it now.